src/refract/RenderFormURLEncodedVisitor.cc

```diff
@@ -0,0 +1,162 @@
+//
+//  RenderFormURLEncodedVisitor.cc
+//  librefract
+//
+//  Created by Kyle Fuller on 17/09/16.
+//  Copyright (c) 2016 Apiary Inc. All rights reserved.
+//
+
+#include <iomanip>
+#include "VisitorUtils.h"
+
+#include "RenderFormURLEncodedVisitor.h"
+
+namespace refract
+{
+
+    namespace {
+        template <typename T>
+        void IncludeMembers(const T& element, typename T::ValueType& members)
+        {
+            const typename T::ValueType* value = GetValue<T>(element);
+
+            if (!value) {
+                return;
+            }
+
+            for (typename T::ValueType::const_iterator it = value->begin();
+                 it != value->end();
+                 ++it) {
+
+                if (!(*it) || (*it)->empty()) {
+                    continue;
+                }
+
+                if ((*it)->element() == "ref") {
+                    HandleRefWhenFetchingMembers<T>(*it, members, IncludeMembers<T>);
+                    continue;
+                }
+
+                members.push_back(*it);
+            }
+        }
+    }
+
+    RenderFormURLEncodedVisitor::RenderFormURLEncodedVisitor() {
+    }
+
+    RenderFormURLEncodedVisitor::~RenderFormURLEncodedVisitor() {
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const IElement& e) {
+        VisitBy(e, *this);
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const MemberElement& element) {
+        if (!result.str().empty()) {
+            result << "&";
+        }
+
+        operator()(*element.value.first);
+
+        result << "=";
+
+        if (element.value.second) {
+            operator()(*element.value.second);
+        } else {
+            result << "";
+        }
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const ObjectElement& element) {
+        if (insideObject) {
+            // Objects inside objects are not supported
+            return;
+        }
+
+        insideObject = true;
+
+        ObjectElement::ValueType value;
+        IncludeMembers(element, value);
+
+        for (std::vector<refract::IElement*>::const_iterator it = value.begin();
+             it != value.end();
+             ++it) {
+
+            if (!*it) {
+                continue;
+            }
+
+            MemberElement *member = static_cast<MemberElement*>(*it);
+            if (member) {
+               operator()(*member);
+            }
+        }
+
+        insideObject = false;
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const EnumElement& element) {
+        const EnumElement::ValueType *value = GetEnumValue(element);
+
+        if (value && !value->empty()) {
+            operator()(*value->front());
+        } else {
+            result << "null";
+        }
+
+    }
+
+    void RenderFormURLEncodedVisitor::operator()(const ArrayElement& e) {
+        // Rendering an array is not supported
```

Comment:
I was looking at how to add a warning for this case but I'm not sure what the best way to be. We wouldn't have source maps here?